async_upnp_client/ssdp.py

```diff
@@ -186,34 +189,60 @@ def _cached_header_parse(
 LOWER_LOCATION = lowerstr("location")
 
 
-def decode_ssdp_packet(
+@lru_cache(maxsize=512)
+def _cached_decode_ssdp_packet(
     data: bytes,
-    local_addr: Optional[AddressTupleVXType],
-    remote_addr: AddressTupleVXType,
+    remote_addr_without_port: AddressTupleVXType,
 ) -> Tuple[str, CaseInsensitiveDict]:
-    """Decode a message."""
+    """Cache decoding SSDP packets."""
     parsed_headers, request_line, udn = _cached_header_parse(data)
     # own data
-    extra: dict[str, Any] = {
-        LOWER__TIMESTAMP: datetime.now(),
-        LOWER__HOST: get_host_string(remote_addr),
-        LOWER__PORT: remote_addr[1],
-        LOWER__LOCAL_ADDR: local_addr,
-        LOWER__REMOTE_ADDR: remote_addr,
-    }
+    extra: dict[str, Any] = {LOWER__HOST: get_host_string(remote_addr_without_port)}
     if udn:
         extra[LOWER__UDN] = udn
 
     # adjust some headers
     location = parsed_headers.get("location", "")
     if location.strip():
         extra[LOWER__LOCATION_ORIGINAL] = location
-        extra[LOWER_LOCATION] = get_adjusted_url(location, remote_addr)
+        extra[LOWER_LOCATION] = get_adjusted_url(location, remote_addr_without_port)
 
     headers = CaseInsensitiveDict(parsed_headers, **extra)
     return request_line, headers
 
 
+def decode_ssdp_packet(
+    data: bytes,
+    local_addr: Optional[AddressTupleVXType],
+    remote_addr: AddressTupleVXType,
+) -> Tuple[str, CaseInsensitiveDict]:
+    """Decode a message."""
+    # We want to use remote_addr_without_port as the cache
+    # key since nothing in _cached_decode_ssdp_packet cares
+    # about the port
+    if len(remote_addr) == 4:
+        if TYPE_CHECKING:
+            remote_addr = cast(AddressTupleV6Type, remote_addr)
+        addr, port, flow, scope = remote_addr
+        remote_addr_without_port: AddressTupleVXType = addr, 0, flow, scope
+    else:
+        if TYPE_CHECKING:
+            remote_addr = cast(AddressTupleV4Type, remote_addr)
+        addr, port = remote_addr
+        remote_addr_without_port = remote_addr[0], 0
+    request_line, headers = _cached_decode_ssdp_packet(data, remote_addr_without_port)
+    return request_line, headers.combine(
+        CaseInsensitiveDict(
+            {
+                LOWER__TIMESTAMP: datetime.now(),
+                LOWER__REMOTE_ADDR: remote_addr,
+                LOWER__PORT: port,
+                LOWER__LOCAL_ADDR: local_addr,
+            }
+        )
+    )
```

Comment:
This is expensive.  Maybe a `.combine_from_lower_dict` will be fast enough since it avoids the object creation